File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    int answer = -1;
    int[] dy = new int[] {-1, 1, 0, 0};
    int[] dx = new int[] {0, 0, -1, 1};
    
    public int solution(int[][] maps) {
        bfs(new int[] {0, 0}, maps);
        
        return answer;
    }
    
    public void bfs(int[] start, int[][] maps){
        int n = maps.length, m = maps[0].length;
        int[] target = new int[] {n-1, m-1};
        int[][] visited = new int[n][m];
        
        Queue<int[]> q = new LinkedList<>();
        
        q.offer(start);
        visited[start[0]][start[1]] = 1;
        
        while (q.size() > 0) {
            int[] now = q.poll();
            int y = now[0];
            int x = now[1];
            
            if (y==target[0] && x == target[1]) {
                answer = visited[y][x];
                break;
            }
            
            for(int dir = 0; dir < 4; dir++) {
                int ny = y + dy[dir];
                int nx = x + dx[dir];
                
                if (ny < 0 || ny >= n || nx < 0 || nx >= m) continue;
                if (maps[ny][nx] == 0) continue;
                if (visited[ny][nx] > 0) continue;
                
                q.offer(new int[] {ny, nx});
                visited[ny][nx] = visited[y][x] + 1;
            }
        }
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    return bfs([0,0], maps);
}

function bfs(start, maps) {
    const [n, m] = [maps.length, maps[0].length];
    const visited = Array.from({length: n}, () => Array(m).fill(0));
    const queue = [start];
    const dy = [-1, 1, 0, 0];
    const dx = [0, 0, -1, 1];
    
    visited[start[0]][start[1]] = 1;
    
    while(queue.length) {
        const [y, x] = queue.shift();
        
        for(let i = 0; i < dy.length; i++) {
            const ny = y + dy[i];
            const nx = x + dx[i];
            
            if(ny >= n || ny < 0 || nx >= m || nx < 0) continue;
            if(maps[ny][nx] === 0) continue;
            if(visited[ny][nx]) continue;
            
            queue.push([ny,nx]);
            visited[ny][nx] = visited[y][x] + 1;
        }
    }
    
    return visited[n-1][m-1] > 0 ? visited[n-1][m-1] : -1;
}